Fixed split-brain merge policy design

Donnerbart opened the pull request:
**Added generics to SplitBrainMergePolicy**

The generics on the `SplitBrainMergePolicy` interface, make it clear which merge types a merge policy requires.

For example the `HigherHitsMergePolicy<V, T extends MergingHits<V>>` requires the `MergingHits` type to be provided by the data structure. This prevents the downcasting to that interface, without knowing if the type is supported.

The merge types can even be composed like `<V, T extends MergingHits<V> & MergingCosts<V>>`, so you can write very specific merge policies in a very simple way.

**Fixed ConfigValidator for generic merge policies**

The merge policy checks in `ConfigValidator` were based on false assumptions regarding `SomeConfig.isStatisticsEnabled()` and valid split-brain merge types. The checks were changed to use reflection and the `SplitBrainMergeTypes` interfaces, which provide the merge types per data structure. The config checks are also also working for custom merge policies now.

The `ConfigValidator` didn't check for invalid split-brain merge policy classes, e.g. `MergePolicyConfig.setPolicy("invalid")`. The checks were changed to allow `SplitBrainMergePolicy` or the according legacy interfaces, depending on the data structure.

**Package overviews**
`com.hazelcast.spi.merge`
![screenshot from 2018-03-27 12-20-43](https://user-images.githubusercontent.com/4196298/37962196-a36d5b22-31ba-11e8-8a89-41d5ffbb95ed.png)
`com.hazelcast.spi.impl.merge`
![screenshot from 2018-03-27 12-23-12](https://user-images.githubusercontent.com/4196298/37962207-ac1d3cf6-31ba-11e8-80d6-cf797bd17a11.png)
`com.hazelcast.internal.config`
![screenshot from 2018-03-27 12-23-37](https://user-images.githubusercontent.com/4196298/37962220-b37d9298-31ba-11e8-8991-d86cc917b7c3.png)

Fixes https://github.com/hazelcast/hazelcast/issues/12284
Fixes https://github.com/hazelcast/hazelcast/issues/12721
Fixes https://github.com/hazelcast/hazelcast/issues/12722
Fixes https://github.com/hazelcast/hazelcast/issues/12725

Donnerbart commented:
run-lab-run

Donnerbart commented on hazelcast/src/main/java/com/hazelcast/internal/config/ConfigValidator.java:
```diff
@@ -65,52 +75,85 @@
+                checkMapMergePolicyWhenStatisticsAreDisabled(mergePolicyClass, requiredMergeTypes);
+            }
+        } catch (InvalidConfigurationException e) {
+            throw new IllegalArgumentException("Merge policy must be an instance of SplitBrainMergePolicy"
```
I changed this for all merge policy related checks, which are in `MergePolicyValidator`. We can decide if we want to change the old checks from `ConfigValidator` as well (but I would do this in another PR).

mmedenjak replied:
👍 I also agree the old checks need to be changed carefully, since we might be breaking compatibility with code relying on older IMDG versions.

Donnerbart replied:
Oh my, we have `InvalidConfigurationException extends RuntimeException` and `ConfigurationException extends HazelcastException`. I'll try to figure out which one is more appropriate later.

Donnerbart commented on hazelcast/src/main/java/com/hazelcast/cache/impl/ICacheRecordStore.java:
```diff
@@ -483,6 +483,15 @@
+     * @param mergePolicy  the {@link SplitBrainMergePolicy} instance to apply
+     * @return the used {@link CacheRecord} if merge is applied, otherwise {@code null}
+     */
+    CacheRecord merge(CacheMergeTypes mergingEntry, SplitBrainMergePolicy<Data, CacheMergeTypes> mergePolicy);
```
My brain doesn't parse this anymore :D `merge(CacheMergeTypes mergingEntry, SplitBrainMergePolicy<Data, CacheMergeTypes> mergePolicy)` is already in the actual code.

I just tried to apply what you've said to `AtomicLong`:
```java
class MergeGreaterValueMergePolicy
    implements SplitBrainMergePolicy<Long, AtomicLongMergingValueImpl> {

    @Override
    public Long merge(AtomicLongMergingValueImpl mergingValue,
                      AtomicLongMergingValueImpl existingValue) {
        if (mergingValue.getValue() > existingValue.getValue()) {
            return mergingValue.getValue();
        }
        return existingValue.getValue();
    }

    @Override
    public void writeData(ObjectDataOutput out) {
    }

    @Override
    public void readData(ObjectDataInput in) {
    }
}
```
This doesn't pass the latest configuration check:
> com.hazelcast.config.ConfigurationException: The merge policy com.hazelcast.concurrent.atomiclong.AtomicLongSplitBrainTest$MergeGreaterValueMergePolicy can just be configured on data structures which provide the merging type com.hazelcast.spi.impl.merge.AtomicLongMergingValueImpl. See SplitBrainMergingTypes for supported merging types.

You can do something like this though:
```java
MergeGreaterValueMergePolicy
    implements SplitBrainMergePolicy<Long, MergingValue<Long>>
```

What am I missing?

mmedenjak replied:
I was thinking on the level of the `ICacheRecordStore#merge` method contract - right now it requires that the parameter must be of type `SplitBrainMergePolicy<Data, CacheMergeTypes> mergePolicy` (for instance `class MergePolicy implements  SplitBrainMergePolicy<Data, CacheMergeTypes>`).
But the method signature doesn't allow covariance in the second generic type - the merge policy can't implement an interface where the second generic type is a subtype - like the one you provided, i.e. this doesn't pass the compiler:
```
SplitBrainMergePolicy<Data, CacheMergingEntryImpl> policy;
cache.merge(mergingEntry, policy);
```

It's the same like this - if you wanted a method that summed all numbers in a collection and wrote:
```
public static int sum(Collection<Number> numbers){
        return ...;
    }
```
This doesn't pass:
```
sum(new ArrayList<Integer>());
        sum(new ArrayList<Float>());
```
So you need to write:
```
public static int sum(Collection<? extends Number> numbers){
        return ...;
    }
```

But never mind, this is actually not necessary and the policy implementing a subtype in the generic parameter doesn't even pass the validation check - which is good. So you can ignore this comment and just keep it in mind for future work with generics.

Donnerbart replied:
A `public CacheRecord merge(CacheMergeTypes mergingEntry, SplitBrainMergePolicy<Data, ? extends CacheMergeTypes> mergePolicy)` breaks some lines later, since (for some reason) `CacheMergeTypes` doesn't match `? extends CacheMergeTypes`.

So we decided to leave it as it for now.

Donnerbart commented:
I just renamed test merge policies from `CustomMapMergePolicyWithoutType` to `CustomMapMergePolicyNoTypeVariable` and adjusted the related test method names etc.

Donnerbart commented:
run-lab-run

Donnerbart commented:
run-lab-run

Donnerbart commented on hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/AbstractEvictableRecordStore.java:
```diff
@@ -325,8 +325,8 @@ protected void mergeRecordExpiration(Record record, EntryView mergingEntry) {
     }
 
     protected void mergeRecordExpiration(Record record, MergingEntry mergingEntry) {
```
Good catch! I knew that there was some downcasting hidden in the code, but just saw it in Cache.

Donnerbart commented on hazelcast/src/main/java/com/hazelcast/map/merge/MergePolicyProvider.java:
```diff
@@ -44,8 +42,7 @@ public MapMergePolicy createNew(String className) {
             } catch (Exception e) {
-                nodeEngine.getLogger(getClass()).severe(e);
-                throw rethrow(e);
+                throw new InvalidConfigurationException("Invalid MapMergePolicy: " + className, e);
```
If there is an exception, it is now first triggered by the `MergePolicyValidator`, then catched and currently rethrown as `ConfigurationException`. For example:
```java
        try {
            Object mergePolicyInstance = mergePolicyProvider.getMergePolicy(mergePolicyClassname);
            checkMergePolicy(mergeTypeProvider, mergePolicyInstance);
        } catch (InvalidConfigurationException e) {
            throw new ConfigurationException("Merge policy must be an instance of SplitBrainMergePolicy"
                    + " or CacheMergePolicy, but was " + mergePolicyClassname);
        }
```
Before that exception should just be triggered during the merge process, which doesn't throw anything at the user. So I think we're free to handle this like we think it's best.

BTW, I'm still unsure about `InvalidConfigurationException` vs. `ConfigurationException`

Donnerbart replied:
I switched to `InvalidConfigurationException`, since it seems to be more frequently used by now (60 vs. 28 constructor calls with this PR), although `ConfigurationException` is older.

Donnerbart commented on hazelcast/src/main/java/com/hazelcast/spi/impl/merge/MergingValueFactory.java:
```diff
@@ -44,35 +54,33 @@
-    public static MergingValue<Data> createMergingValue(SerializationService serializationService, CollectionItem item) {
-        return new MergingValueImpl<Data>(serializationService)
+    public static CollectionMergeTypes createMergingValue(SerializationService serializationService, CollectionItem item) {
+        return (CollectionMergingValueImpl) new CollectionMergingValueImpl(serializationService)
```
Oh, yes, that was an temporary fix, before each class got it's own wrapper class. Good catch.

Donnerbart commented on hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractHazelcastCacheManager.java:
```diff
@@ -114,9 +114,8 @@ public AbstractHazelcastCacheManager(CachingProvider cachingProvider, HazelcastI
-        checkCacheConfig(newCacheConfig.getInMemoryFormat(), newCacheConfig.getEvictionConfig(),
-                newCacheConfig.isStatisticsEnabled(), newCacheConfig.getMergePolicy());
-
+        // FIXME: we should pass the CacheMergePolicyProvider here
```
Good reminder. I'm not sure if Caches created via `HazelcastCacheManager` are part of the split-brain handling at all. I didn't see them being added to the cache service, which is being looked up during the split-brain healing. I'll check this with Vassilis.

Donnerbart replied:
(I would also have to get my hands onto the `CacheMergePolicyProvider` from the given `HazelcastInstance`, which can be a client or server -> that's gonna be fun).

vbekiaris replied:
> I'm not sure if Caches created via HazelcastCacheManager are part of the split-brain handling at all. I didn't see them being added to the cache service, which is being looked up during the split-brain healing. 

They are handled, split-brain handler cache service will handle on the basis of `CacheConfig`s registered in the `CacheService.configs` and existing partition segments.

In terms of actually fixing the `FIXME`, the check should be delegated to the `HazelcastServerCacheManager` or `HazelcastClientCacheManager` to differentiate accordingly:
- `HazelcastServerCacheManager` has a reference to `CacheService` -> can obtain and use the `CacheMergePolicyProvider`.
- `HazelcastClientCacheManager` may have to exclude the merge policy validation on the client side at that point and perform the validation on the member side. This is what happens next, anyway, `createCacheConfig` a couple of lines underneath sends a request to the member side which is handled in `CacheCreateConfigMessageTask`, merge policy validation could happen there.

Donnerbart replied:
Thanks for the hints! I've added the checks to both places
* `abstract void AbstractHazelcastCacheManager.validateCacheConfig(CacheConfig<K, V> cacheConfig)`, which does the limited check on the client side, and the full one on the member side
* `CacheCreateConfigMessageTask.call()` with a full config check

I also added `MergePolicyValidatorCachingProviderIntegrationTest` to test the config checks when `HazelcastCachingProvider` is used. Works as expected, so no surprises here.

vbekiaris replied:
Looks good! 👍
